[1.20.4] Stack aware `getFoodComponent`

modmuss50 opened the pull request:
99% of the work was done in #3295 by @Phoupraw just a 1.20.4 version

~~This is the same, just targgeting 1.20.4 and using @WrapOperation in place of @Redirect~~

modmuss50 committed:
Stack aware `getFoodComponent`

Co-authored-by: Phoupraw <50520903+phoupraw@users.noreply.github.com>

LlamaLad7 commented:
I wouldn’t use WrapOperation for any of those. They are all meaningful unconditional replacements and as such they should always be the innermost thing in a wrapper chain, and if another mod wants to do a meaningful unconditional replacement as well then they should conflict. For those reasons Redirect is the right choice. The ThreadLocal should however be replaced with a `@Share`.

modmuss50 committed:
Back to redirects

modmuss50 commented:
> I wouldn’t use WrapOperation for any of those. They are all meaningful unconditional replacements and as such they should always be the innermost thing in a wrapper chain, and if another mod wants to do a meaningful unconditional replacement as well then they should conflict. For those reasons Redirect is the right choice. The ThreadLocal should however be replaced with a `@Share`.

Reverted that change and moved to share. Thats a good point, I didnt really give it much thought 👍

LlamaLad7 reviewed (approved):
LGTM. There are situations in which you would want to wrap even without calling the original, but as with everything in mixin it’s nuanced.
